iceoryx_posh/include/iceoryx_posh/internal/popo/client_impl.inl

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2021 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_POSH_POPO_CLIENT_IMPL_INL
+#define IOX_POSH_POPO_CLIENT_IMPL_INL
+
+#include "iceoryx_posh/internal/popo/client_impl.hpp"
+
+namespace iox
+{
+namespace popo
+{
+template <typename Req, typename Res, typename BaseClientT>
+ClientImpl<Req, Res, BaseClientT>::ClientImpl(const capro::ServiceDescription& service,
+                                              const ClientOptions& clientOptions) noexcept
+    : BaseClientT(service, clientOptions)
+{
+}
+
+template <typename Req, typename Res, typename BaseClientT>
+cxx::expected<Request<Req>, AllocationError> ClientImpl<Req, Res, BaseClientT>::loanUninitialized() noexcept
+{
+    auto result = port().allocateRequest(sizeof(Req), alignof(Req));
+    if (result.has_error())
+    {
+        return cxx::error<AllocationError>(result.get_error());
+    }
+    else
+    {
+        auto requestHeader = result.value();
+        auto payload = mepoo::ChunkHeader::fromUserHeader(requestHeader)->userPayload();
+        auto request = cxx::unique_ptr<Req>(reinterpret_cast<Req*>(payload), m_requestDeleter);
+        return cxx::success<Request<Req>>(Request<Req>{std::move(request), *this});
+    }
+}
+
+template <typename Req, typename Res, typename BaseClientT>
+template <typename... Args>
+cxx::expected<Request<Req>, AllocationError> ClientImpl<Req, Res, BaseClientT>::loan(Args&&... args) noexcept
+{
+    return std::move(
+        loanUninitialized().and_then([&](auto& request) { new (request.get()) Req(std::forward<Args>(args)...); }));
+}
+
+template <typename Req, typename Res, typename BaseClientT>
+void ClientImpl<Req, Res, BaseClientT>::send(Request<Req>&& request) noexcept
+{
+    auto payload = request.release(); // release the Request ownership of the chunk before publishing
```

Comment:
Here it is maybe interesting why are you doing this before and not after `sendRequest` otherwise I find this comment a little bit obvious.

Reply:
Because `sendRequest` takes the ownership and if I don't release if from the `Request` if would also hold it. This might be a case where `cxx::unique_ptr` would be more appropriate to communicate the ownership transfer to `sendRequest`. Currently all the ports use raw pointer. Shall I open an issue to clean this up for all ports?

Reply:
I already updated the comment to emphasis more the why instead of the what